src/adapter/graphql/src/mutations/dataset_mut/adapters_mut/collection_mut.rs

```diff
@@ -0,0 +1,427 @@
+                    .map(|entry| (entry.path.clone(), entry))
+                    .collect()
+            }
+        };
+
+        let mut diff = Vec::new();
+
+        for op in operations {
+            if let Some(add) = op.add {
+                if let Some(existing) = current_entries.remove(&add.entry.path) {
+                    diff.push((Op::Retract, existing));
+                }
+
+                let new_entry = CollectionEntry::from_input(dataset.clone(), add.entry);
+                current_entries.insert(new_entry.path.clone(), new_entry.clone());
+                diff.push((Op::Append, new_entry));
+            } else if let Some(remove) = op.remove {
+                if let Some(existing) = current_entries.remove(&remove.path) {
+                    diff.push((Op::Retract, existing));
+                }
+            } else if let Some(mov) = op.r#move {
+                let Some(old_entry) = current_entries.remove(&mov.path_from) else {
+                    return Ok(CollectionUpdateResult::NotFound(
+                        CollectionUpdateErrorNotFound {
+                            path: mov.path_from,
+                        },
+                    ));
+                };
+
+                let mut new_entry = old_entry.clone();
+                new_entry.path = mov.path_to;
+                if let Some(extra_data) = mov.extra_data {
+                    new_entry.extra_data = extra_data;
+                }
+
+                if new_entry.path != old_entry.path {
+                    if let Some(collision) = current_entries.remove(&new_entry.path) {
+                        diff.push((Op::Retract, collision));
+                    }
+
+                    diff.push((Op::Retract, old_entry));
+                    diff.push((Op::Append, new_entry.clone()));
+                } else {
+                    diff.push((Op::CorrectFrom, old_entry));
+                    diff.push((Op::CorrectTo, new_entry.clone()));
+                }
+
+                current_entries.insert(new_entry.path.clone(), new_entry);
+            }
+        }
+
+        if diff.is_empty() {
+            return Ok(CollectionUpdateResult::UpToDate(CollectionUpdateUpToDate));
+        }
+
+        let mut expected_head: odf::Multihash = expected_head;
+        let mut res = None;
+
+        // TODO: PERF: FIXMEEEEEEEEEEEEEEEEEEEEEEEEEEEEE: Writing each operation in a
+        // different block to work around changelog sorting issue. This is a performance
+        // killer.
+        for op in diff {
+            match self
+                .write_records(ctx, vec![op], Some(expected_head.clone()))
+                .await?
+            {
+                CollectionUpdateResult::Success(r) => {
+                    expected_head = r.new_head.clone().into();
+                    res = Some(CollectionUpdateResult::Success(r));
+                }
+                r => {
+                    return Ok(r);
+                }
+            }
+        }
```

Comment:
We rely on sorting by `(pk, op)` in merge strategies, but this fails for changelog streams that can contain multiple retractions/corrections for the same key within a single data batch. I will ticket up and reference this problem here, but this workaround of splitting changes into multiple blocks should last us for a while.